Project Proposal: Openverse Datasets

zackkrida opened the pull request:
<!-- prettier-ignore -->
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->

Related to #2545

## Description
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->

This PR adds a project proposal for the Dataset project. I've tried to get this out quickly so it might be a bit rough. Suggestions are very welcome. I've asked @sarayourfriend (for providing past feedback on this initiative) and @AetherUnbound (for general data expertise) to review from the Openverse side.

I would also appreciate insights from @apolinario on the HuggingFace platform: how it relates to _this project_ but also some of its general functionality which I touch on in the proposal.

## Descisionmaking

This discussion is following the Openverse decision-making process. Information about this process can be found [on the Openverse documentation site](https://docs.openverse.org/meta/decision_making/index.html).

Requested reviewers or participants will be following this process. If you are being asked to give input on a specific detail, you do not need to familiarise yourself with the process and follow it.

### Current round

This discussion is currently in the **[Decision round](https://docs.openverse.org/meta/decision_making/process_description.html#revision-round)**.

Will be resolved by **2023-07-20**.

## Testing Instructions
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->

Read the document in GitHub's code view or the [generated docs preview](https://docs.openverse.org/_preview/2637/projects/proposals/publish_dataset/20230706-project_proposal.html).

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [x] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

apolinario commented on documentation/projects/proposals/publish_dataset/20230706-project_proposal.md:
```diff
@@ -0,0 +1,207 @@
+  [Vision Transformer](https://huggingface.co/google/vit-base-patch16-224)
+- [Duplicate image detection #427](https://github.com/WordPress/openverse/issues/427) -
+  Deduplication is an extremely common preprocessing step. We should be able to
+  apply the deduplication from
```
There are a few open source matching image techniques - including if they are cropped or of different quality. I was chatting with @Skylion007 on some of those techniques, including [sscd](https://github.com/facebookresearch/sscd-copy-detection), which can be a promising way

This is something that, once the dataset is out, could be contributed upstream to the database, with a column `is_duplicate_to` or similar

Skylion007 replied:
I think this is very important as there are likely to be ALOT of duplicate images here, especially between image sources (ie. I expect NASA images to appear on Flickr in various organizations as well as wikimedia etc).

apolinario commented on documentation/projects/proposals/publish_dataset/20230706-project_proposal.md:
```diff
@@ -0,0 +1,207 @@
+- Develop a plan for updating the datasets regularly
+- Refine how we will provide access to our raw dataset for upload to
+  HuggingFace:
+  - Delivery mechanism (a requester pays S3 bucket?)
```
On HF side we don't plan on having bottlenecks or paid egress for the dataset. It would be fully free to access and download the entire dataset or parts of it via streaming. 

If the final dump size in the final format doesn't fit our traditional datasets repos - we can provide infra for it to be always openly accessible

Skylion007 replied:
At least on Mosaic's end, happy to do a requester pays bucket or copy the data to another S3 bucket in the same region and then we can copy to one of our OCI buckets

apolinario commented on documentation/projects/proposals/publish_dataset/20230706-project_proposal.md:
```diff
@@ -0,0 +1,207 @@
+- Refine how we will provide access to our raw dataset for upload to
+  HuggingFace:
+  - Delivery mechanism (a requester pays S3 bucket?)
+  - File format (parquet files? TSVs? a Postgres dump?)
```
I think the HF Datasets preferred format are `parquet` files, although more are compatible with it. We would recommend having it converted to parquet and split into smaller parquet files for streamability and ease to consume (plus, `parquet` enables all features of HF Datasets + some extra external features such as [DuckDB compatibility](https://huggingface.co/blog/hub-duckdb))

zackkrida replied:
Clarified that this refers to the initial dump only in 5476f48ef

apolinario commented on documentation/projects/proposals/publish_dataset/20230706-project_proposal.md:
```diff
@@ -0,0 +1,207 @@
+it:
+
+- Reach out to past requesters of the dataset and share the HuggingFace link
+- Social channel cross-promotion between the WordPress Marketing team,
```
Happy to do cross comms here 🤗

Skylion007 replied:
Also happy to do cross-comms as well.

zackkrida replied:
Great, thank you both! We'll get you in touch with the WordPress project's marketing team in the future to help coordinate.

apolinario commented on documentation/projects/proposals/publish_dataset/20230706-project_proposal.md:
```diff
@@ -0,0 +1,207 @@
+<!-- What are the required implementation plans? Consider if they should be split per level of the stack or per feature. -->
+
+- Initial Data Dump Creation - A plan describing how to produce and provide
+  access to the raw data dumps which will be used to create the Dataset(s).
```
Happy to help on the conversion to the HF Datasets format part of this process. Of course there's an internal pre-processing until it gets there, but if you would like to provide a raw dump at some point I'd be happy to work with you on finding the best format from that raw dump to something publishable

Skylion007 commented on documentation/projects/proposals/publish_dataset/20230706-project_proposal.md:
```diff
@@ -0,0 +1,207 @@
+- Refine how we will provide access to our raw dataset for upload to
+  HuggingFace:
+  - Delivery mechanism (a requester pays S3 bucket?)
+  - File format (parquet files? TSVs? a Postgres dump?)
```
There are two file format questions here, the initial data dump used to download all the images before we upload to HF and the later dataset we will actually host. The former does not really matter as we will parse or transform any data you give us. For the image data, there are ton of formats we can pick including webdataset, parquet, mosaicml datashards, etc.

zackkrida replied:
Clarified that this refers to the initial dump only in 5476f48ef

apolinario commented:
Heya, I think the project proposal looks great @zackkrida! Thanks for putting it together! I made very minor comments across the doc

> how it relates to this project but also some of its general functionality which I touch on in the proposal. 

I think the document has a fair summary of the capabilities of the platform relative to this project. I would just add that the [streaming](https://huggingface.co/docs/datasets/stream) feature of datasets allow for making it accessible to people that may not have access to storage that allows storing the entire dataset. May be helpful as part of the democratisation of this data

Skylion007 commented:
> Heya, I think the project proposal looks great @zackkrida! Thanks for putting it together! I made very minor comments across the doc
> 
> > how it relates to this project but also some of its general functionality which I touch on in the proposal.
> 
> I think the document has a fair summary of the capabilities of the platform relative to this project. I would just add that the [streaming](https://huggingface.co/docs/datasets/stream) feature of datasets allow for making it accessible to people that may not have access to storage that allows storing the entire dataset. May be helpful as part of the democratisation of this data

+1 on streaming the dataset. It also can allow people to easily and quickly generate various subsets of the data.

AetherUnbound commented on documentation/projects/proposals/publish_dataset/20230706-project_proposal.md:
```diff
@@ -0,0 +1,207 @@
+## Project summary
+
+This project aims to publish and regularly update a dataset or _datasets_ of
+Openverse' media metadata. This project will provide access to the data
```
Should this be?
```suggestion
Openverse's media metadata. This project will provide access to the data
```

AetherUnbound commented on documentation/projects/proposals/publish_dataset/20230706-project_proposal.md:
```diff
@@ -0,0 +1,207 @@
+
+This project aims to publish and regularly update a dataset or _datasets_ of
+Openverse' media metadata. This project will provide access to the data
+currently served by our API, but which is difficult to access in full, requires
+significant time, money, and compute resources.
```
Maybe?
```suggestion
currently served by our API, but which is difficult to access in full and requires
significant time, money, and compute resources to maintain.
```

AetherUnbound commented on documentation/projects/proposals/publish_dataset/20230706-project_proposal.md:
```diff
@@ -0,0 +1,207 @@
+for all users. It also requires significant maintainer effort to identify,
+mitigate, and block scraping traffic.
+
+By sharing this data as a full dataset on
+[HuggingFace](https://huggingface.co/datasets), we can remove these barriers
+allow folks to access the data provided by Openverse.org and the Openverse API
```
```suggestion
By sharing this data as a full dataset on
[HuggingFace](https://huggingface.co/datasets), we can remove these barriers and
allow folks to access the data provided by Openverse.org and the Openverse API
```

AetherUnbound commented on documentation/projects/proposals/publish_dataset/20230706-project_proposal.md:
```diff
@@ -0,0 +1,207 @@
+[data preprocessing](https://huggingface.co/docs/transformers/preprocessing)
+stage of model training.
+
+HuggingFace in particular will enable community members to analyze the dataset
```
This wording feels to me like this is an obligation HuggingFace has after the fact. Would you be comfortable with this slight rewarding so it better conveys a sense that this is what HuggingFace is currently capable of?

```suggestion
Presence on HuggingFace in particular will enable community members to analyze the dataset
```

AetherUnbound commented on documentation/projects/proposals/publish_dataset/20230706-project_proposal.md:
```diff
@@ -0,0 +1,207 @@
+  [Vision Transformer](https://huggingface.co/google/vit-base-patch16-224)
+- [Duplicate image detection #427](https://github.com/WordPress/openverse/issues/427) -
+  Deduplication is an extremely common preprocessing step. We should be able to
+  apply the deduplication from
```
It looks like there was something else that was intended to be added to this sentence, is that the case?

zackkrida replied:
Fixed in 38de4678b

AetherUnbound reviewed (approved):
This project plan looks excellent, thank you for drafting it Zack! My notes are merely surface level/wording, I'm aligned with everything else here 🙂

sarayourfriend reviewed (approved):
This looks great. I'd like clarification and/or additional information about the following points:

- Licensors, from my perspective, are also stakeholders. Respecting their intentions and properly communicating the usage conditions is especially important for a project where every single work has an explicit license. Noting specific license elements like NC, ND, and SA that have known nuances in the distribution of the dataset feels important enough to list as a requirement.
- On the other hand, "licensors" are not the only stakeholders from that perspective and PDM works present a significant complication in this regard, both from a regional legal perspective and from the perspective of the communication of cultural artefacts that the institutions distributing PDM marked works based on those artefacts have obtained and distributed without consultation or otherwise involvement of the culture the artefacts were taken from. Openverse can't do much to fix the underlying problems, but we do need to take care to protect our liability in this regard the same way we do in [our general terms of service](https://docs.openverse.org/terms_of_service.html#content-and-data-available-through-the-api). Should the dataset have the same terms of service applied? A specific terms of service/disclaimer worded directly for the dataset would better protect the project, I imagine.
- Does the first implementation plan also include the documentation updates you mentioned? Can that be listed as an explicit requirement? It is easy to miss something like that when writing an implementation plan that is sure to already be significant in other respects.

Anyway, everything sounds good to me. The rationale to use HuggingFace makes sense. My only concern moving forward is to ensure that we've covered our bases as far as liability to ourselves and have communicated as effectively as possible to users of the dataset their responsibility in using the dataset.

Nothing else to add on top of what others have shared.

zackkrida commented:
Drafting this proposal while I move it into the [Revision round](https://docs.openverse.org/meta/decision_making/process_description.html#revision-round).

Skylion007 commented on documentation/projects/proposals/publish_dataset/20230706-project_proposal.md:
```diff
@@ -0,0 +1,213 @@
+- Creative Commons - Stewards of the Commons and CC Licenses, advisors, and
+  another partner in marketing promotion
+- Aaron Ghokaslan - A researcher working on supplementary datasets and providing
+  technical advice
```
MosaicML is happy to provide help with marketing collaboration as well. :)

zackkrida commented:
I've addressed reviewer comments and this proposal is now ready for a [decision](https://docs.openverse.org/meta/decision_making/process_description.html#decision-round).

sarayourfriend reviewed (approved):
LGTM! Shall we add a line item to the priorities meeting to discuss the implementation plans? How does this fall in line with the rest of our work? (we can discuss this on the project thread, if it's more involved)

sarayourfriend commented on documentation/projects/proposals/publish_dataset/20230706-project_proposal.md:
```diff
@@ -0,0 +1,225 @@
+
+<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->
+
+- [ ] @AetherUnbound
+- [ ] @sarayourfriend
```
```suggestion
- [x] @sarayourfriend
```

zackkrida replied:
```suggestion
- [x] @AetherUnbound
- [x] @sarayourfriend
```

zackkrida commented:
I'll leave some scheduling thoughts on the project thread, @sarayourfriend

zackkrida commented:
@apolinario yesterday I saw that the https://huggingface.co/meta-llama/Llama-2-7b model has an access flow which requires accepting terms and signing up through a meta controlled page: https://ai.meta.com/resources/models-and-libraries/llama-downloads/

I am curious: is this type of flow available to _datasets_, as well? We wouldn't necessarily want folks to wait 1-2 days for access, but the idea of more-explicitly enforcing our terms for dataset usage (respecting proper usage and whatnot) is appealing.

sarayourfriend commented:
@zackkrida it is possible for datasets: https://huggingface.co/docs/hub/datasets-gated

apolinario commented:
As @sarayourfriend noted, gating repos is totally available for datasets! 

And it comes in two flavours: `manual approval` which basically restricts who filled in the form and has the approval done manually (or automated that via a Django service). 

BUT if you are going to automate the approvals, the other mode of gated access (`automatic approval`) would make more sense imo. Basically this requires people to read/accept information they can use it, but everyone that does is accepted automatically

It is what Mozilla uses for common voice (example [here](https://huggingface.co/datasets/mozilla-foundation/common_voice_4_0)) - I think that could make sense for this project as well 

<img width="401" alt="image" src="https://github.com/WordPress/openverse/assets/788417/d220c5b8-6649-459d-9c4b-d5849a35dbfe">

zackkrida commented:
Thanks, @apolinario! I am going to draft this proposal while we make more efforts to consider our specific use conditions and ethical standards for the dataset, along with how they would relate to the "Gating" functionality.

apolinario commented:
Sounds good! Here's an idea that I had that I hope could help tackling a few challenges with appropriate disclaimer, and downstream usage concerns 

### Multiple-subsets
Instead of a single Openverse dataset (or two, one for visual media and one for audio), we could create multiple subsets based on license or license-grouping, e.g. (not really name suggestion heres, just examples): 
- `openverse/images-cc-by`
- `openverse/images-cc-by-nc`
- `openverse/images-cc0`
etc.

All in the same format, but each could have its own dataset card and its own set of disclaimers and descriptions (and all under your terms of service disclaimers ofc). In one hand, this could make using this datasets for downstream tasks a bit more convoluted/complex (as now one has to engage with/process multiple datasets), on the other hand, it would make it very obvious what each dataset could be for, and it could inform downstream users very specifically what they are doing, as they would need to write code that looks like: 
```py
from datasets import load_dataset
dataset_cc_by = load_dataset("openverse/images-cc-by", token=True)
dataset_cc_by_nc = load_dataset("openverse/images-cc-by-nc", token=True)
#do downstream tasks processing both
```
That could make it pretty clear that they should **not** do that if they are looking into a commercial application - while a single dataset we would need to require the user to apply some column filtering in case of downstream applications. Which is easy enough with the HF `datasets` library, but maybe this is a way to make it even more explicit and understood even from reading the code - and not only by reading the model card. 

This could also co-exist with gating, so each dataset repo could be gated. Btw the `load_dataset` function gives a 403 error if the dataset is gated and your HF user isn't in.

krysal commented on documentation/projects/proposals/publish_dataset/20230706-project_proposal.md:
```diff
@@ -0,0 +1,225 @@
+or made easier by the publication of the datasets. This could work in a few
+ways. A community member, training a model using the Openverse dataset,
+generates metadata that we want and planned to generate ourselves. Then, the
+HuggingFace platform presents an alternative to other SaSS products we intended
```
Is "SaSS" the short version of something? I can't find a different meaning other than the CSS extension language, [SASS](https://sass-lang.com/). Can we add the full form or the meaning in a note/footnote, maybe?

zackkrida replied:
typo for Saas (software as a service), I'll update to explain.

krysal commented on documentation/projects/proposals/publish_dataset/20230706-project_proposal.md:
```diff
@@ -0,0 +1,225 @@
+This project encompasses all of our 2023 "lighthouse goals", but "Community
+Development" is perhaps the broadest and most relevant. Others touched on here,
+or impacted through potential downstream changes, are "Result Relevancy",
+"Quantifying our Work","Search Experience", "Content Safety", and "Data
+inertia".
```
```suggestion
"Quantifying our Work", "Search Experience", "Content Safety", and "Data
inertia."
```

AetherUnbound replied:
Corrected the spacing, but the period should remain outside the quote here.

krysal reviewed (approved):
I love the several good points captured here. Just restating what has been said, but it's an excellent proposal!

AetherUnbound commented:
Merging this for now so the document is in our documentation site, even though we are [not planning on pursuing it at this time](https://github.com/WordPress/openverse/issues/2545#issuecomment-1682627814).
